### ISSUE TITLE
How to using RepGraph to a BP project by myself  ?

### ISSUE THREAD

**Issue by xiaoaq** [locus84/LocusReplicationGraph#2]
Hi, Locus, I'm sorry to disturb you again.
Should I using BasicReplicationGraph.h if my project base classes are not in C++, since I can't set routings unless the bases are in c++﻿ , right?
And how much percentage can be optimize if only BasicReplicationGraph is used in BP project ?﻿

**Comment by locus84** [comment]
It doesn't matter your base class is BP or not. you can set whatever you want in the LocusReplicationGraph inherited blueprint.
![image](https://user-images.githubusercontent.com/6591432/53858613-41e16d80-401e-11e9-9145-77acf4d20aa2.png)

But your project should not be BP only because plugin itself is c++

**Comment by xiaoaq** [comment]
Do I need to care about switching between different levels(or worlds)?
If not, I can add all the classes that need to be added in level to ReplicationPolicySettings/ReplicationInfoSettings at once?

**Comment by locus84** [comment]
No you don't need to care about scenes. It's bound to NetDriver so the setting will be valid across network scenes. 
Yes, just add your classes there.

**Comment by xiaoaq** [comment]
Got it! 
If the ReplicationPolicySettings/ReplicationInfoSettings is not set up, the plugin implements the basic functions, like BasicReplicationGraph, right? If so, will it be significantly optimized?

**Comment by locus84** [comment]
It depends, But I strongly recommend you to tweak those settings as low as possible, and expend further later on when needed.

**Comment by xiaoaq** [comment]
When GetChecked my own class, will throw exception, like 'No ClassInfo found for BP_A_Gun_C'.
`const bool bClassIsSpatialized = IsSpatialized(ClassRepNodePolicies.GetChecked(ReplicatedClass));`
![image](https://user-images.githubusercontent.com/33898493/53928010-925ed680-40c3-11e9-9aba-9e48c42e9273.png)

**Comment by xiaoaq** [comment]
Exceptions will also be thrown after running for a period of time.

![image](https://user-images.githubusercontent.com/33898493/53931599-d6a4a380-40d0-11e9-87ed-c900f11c483d.png)

![image](https://user-images.githubusercontent.com/33898493/53931619-ef14be00-40d0-11e9-9d12-e1726cd44dd4.png)

I'm not sure what caused it..

**Comment by locus84** [comment]
Thank you for sharing these errors, can you share me callstacks? so I can look into it.
Have you added more than 200ish items in a grid?

**Comment by locus84** [comment]
You have to list up all networked actors in Replication Info Settings.
At least it's root AActor class.

**Comment by xiaoaq** [comment]
Sure!
This is the callstacks for question which exception at GetChecked .
![image](https://user-images.githubusercontent.com/33898493/53945417-8cd1b280-40fc-11e9-819d-4682ef06b81d.png)

**Comment by xiaoaq** [comment]
Is there a limit on the number of items per gird?
This is the level I tested, and there will be AI generated after Playing.
![image](https://user-images.githubusercontent.com/33898493/53945801-619b9300-40fd-11e9-92d9-e029530b8177.png)
___________________________________________________________________________________________________________
BTW，I'm sorry If I don't understand your means very well. In fact, I'm not good at English. Plz forgive me. ^ ^

**Comment by xiaoaq** [comment]
> You have to list up all networked actors in Replication Info Settings.
> At least it's root AActor class.
I didn't adjust the plugin setting. I used the default settings. Is that the problem?

**Comment by locus84** [comment]
You have to add your gun BP to the setting list

**Comment by xiaoaq** [comment]
Well, I seem to understand.
Take the gun for example, I should add the gun to Replication Policy Settings and set it to NotRouted; then when Character picks up the gun, I add the gun to Character through the AddDependent Actor, and synchronize with Character.
As for Character, I need to set not only Policy, but also ReplicationInfo.
Is that so?

**Comment by locus84** [comment]
Exactly, because there's already a setting for Pawn which is parent class of Character. You don't have to.
Likewise, if there's a parent class that represent all of you item classes(let's say Gun is child class of Item), you can add it into Info settings, leads all of child class(Gun, for example) will follow that that setting.

**Comment by xiaoaq** [comment]
Got it!
So if I don't check 'Include Child Class', even if a class inherits it, the child class won't follow its settings, right?
Hypothesis，Gun is a child-class of Character and bullet is a child-class of gun(via bullet AddDependentActor to gun?). Will bullets be follow Character's settings?
------------------------------------------------------------------------------
I also have a question: Pawn set RepInfo but not set RepPolicy at the same time? So what knows what kind of policy it is?

**Comment by locus84** [comment]
Default RepPolicy for all networked actors are 'Spatialized Dynamic'. If you want 'Always relevant' or 'Spatialized static' or 'Not routed'(like your Gun), you need add your own policy for that class.
You can use LocusRepGraph.PrintRouting console command to check actors' policy.

**Comment by xiaoaq** [comment]
Thank you very much for your patient and detailed answer! I basically understand this part.
————————————————————————————————————————

![image](https://user-images.githubusercontent.com/33898493/54018690-519ab680-41c4-11e9-834c-083d2b41943c.png)
About the params of RepGraph & RepClassInfo, I know that adjustments need to be made according to actual projects, but is there have a basic rule can follow ?

In my test scene, this error is always reported, and I tried to adjust some parameters.
I reduced the cell size, increased the number of buckets, but no use also? Am I putting too many items in a small range?

![image](https://user-images.githubusercontent.com/33898493/54018390-7e020300-41c3-11e9-8d5a-a6566007927d.png)
![image](https://user-images.githubusercontent.com/33898493/54018439-9bcf6800-41c3-11e9-904c-4219dfc15d86.png)

**Comment by xiaoaq** [comment]
And, what is the role of Buckets?

**Comment by locus84** [comment]
Exactly, you're putting too many Items in small range. That error is came from pure ReplicationGraph plugin, not from mine. How's put your non-player actor to Spatialized static policy?

If you're doing some kind of stress test, you better scatter your objects into larger area.
And reduce cull distance squard(to reduce view range).

**Comment by locus84** [comment]
Bucket is kind of a list which holds actors that should be relevant to an actor in corresponding grid.
The error means that requested bucket size is too large to allocate.

**Comment by xiaoaq** [comment]
Yes, that's the problem. 
How should I setting the rep-policy for the actor spawn in the playing?
If static existence is known after spawn, set it Spatialize_Static; but if the actor can be destroyed, can it be set to Spatialize_Dormancy? If it's dormancy policy, do I need to manage the actor's dormancy attribute after setting it up?
